### DISCUSSION TITLE
EnumArrayType is broken

### DISCUSSION THREAD

**Discussion by TrejGun** [mikro-orm/mikro-orm#2084] [Q&A]
**Describe the bug**
Can make enum[] work in postgress

**Stack trace**
```
[Nest] 44183  - 07/31/2021, 2:11:22 AM   ERROR [ExceptionsHandler] Invalid enum array items provided in UserEntity.roles: [ '{USER}' ]
ValidationError: Invalid enum array items provided in UserEntity.roles: [ '{USER}' ]
    at EnumArrayType.convertToDatabaseValue (~/node_modules/@mikro-orm/core/types/EnumArrayType.js:26:23)
    at ~/node_modules/@mikro-orm/core/hydration/ObjectHydrator.js:72:96
    at eval (eval at createFunction (~/node_modules/@mikro-orm/core/utils/Utils.js:611:52), <anonymous>:12:20)
    at Function.callCompiledFunction (~/node_modules/@mikro-orm/core/utils/Utils.js:622:20)
    at ObjectHydrator.hydrate (~/node_modules/@mikro-orm/core/hydration/ObjectHydrator.js:24:23)
    at EntityFactory.hydrate (~/node_modules/@mikro-orm/core/entity/EntityFactory.js:82:27)
    at EntityFactory.create (~/node_modules/@mikro-orm/core/entity/EntityFactory.js:36:14)
    at SqlEntityManager.findOne (~/node_modules/@mikro-orm/core/EntityManager.js:242:42)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at AuthService.login (~/apps/mikro-orm/src/auth/auth.service.ts:39:24)
    at ~/node_modules/@nestjs/core/router/router-execution-context.js:46:28
    at ~/node_modules/@nestjs/core/router/router-proxy.js:9:17
```

**To Reproduce**
1. create entity with enum[] field
```ts
  @Enum({ items: () => UserRole, array: true, default: [UserRole.USER] })
  public roles: Array<UserRole>;
```
2. try to read or save
3. check console log

**Expected behavior**
userEntity.roles should be an array of strings

**Additional context**
I tried to print what is inside EnumArrayType::convertToDatabaseValue
and it appears to be string '{USER}' rather than an array of strings ['USER']

you can find an example here 
https://github.com/TrejGun/nestjs-database
follow instructions in the readme or contact me in telegram by my nickname

**Versions**

| Dependency | Version |
| - | - |
| node | v14.17.0 |
| typescript | 4.3.5 |
| mikro-orm | 4.5.7 |
| your-driver | postgres:12 |

**Comment by B4nan** [comment]
Please provide reproduction _only_ for this problem, your reprository containts too many code. Also don't use any other framework, especially unsupported versions of nest.

> try to read or save

This is exactly where you should say what you are doing and provide the data too. Guess what, we have this covered in tests and it is working just fine, e.g. here:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/issues/GH482.test.ts#L35

Ideally reproduction should be a single file test case, like this one, I can't browse your huge repository with hundreds of files and many unneeded build steps or tool configs that are totally unrelated. Reproduction != example app.

**Reply by TrejGun** [reply]
this is all nice and I try to check out your repo and run your tests but your MIKRO-orm is about 1.5Gb and requires installing python which requires installing XCode 11Gb zipped and then docker with progress. sorry guys this just does not with it. 

This bug report was my contribution to open source, it is small but I can't afford more. So you are aware.

**Reply by B4nan** [reply]
You don't need xcode nor python, you need just command line tools (= basically git).

**Reply by TrejGun** [reply]
~~While I was trying to set up your tests I noticed your docker-compose uses Postgres 13 and my example was using 12 so I updated the DB version and my code just works now~~

oops I just ran wrong example. it still fails

**Reply by TrejGun** [reply]
```sql
create type user_role_enum as enum (
      'ADMIN',
      'USER'
);
      
create table user (
      id serial primary key,
      roles user_role_enum[] not null default '{USER}'
);
```

is this SQL correct? I create DB structure manually and it might be different from shat you generate from models

**Reply by B4nan** [reply]
Nope, MikroORM uses knex under the hood and enums are done as text colums with check constraints. For enum arrays, instead of using check constraint, we use runtime validation (that is the place throws the error from your OP). Native postgres enums are currently not supported.

**Reply by TrejGun** [reply]
ok here is what you do instead
```
create table "mikroorm"."user" (
"id" serial primary key, 
"roles" text[] not null default '{USER}', 
"status" text check ("status" in ('ACTIVE', 'INACTIVE', 'PENDING')) not null default 'PENDING'
); 
```
well... EnumArrayType works fine.

according to this thread
https://github.com/mikro-orm/mikro-orm/issues/476
it is possible to create something like NativeEnumArrayType for my needs
